### PR TITLE
*: enable gocritic.exitAfterDefer

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve?

Issue Number: Ref https://github.com/tikv/pd/issues/4322

ref https://www.quasilyte.dev/blog/post/log-fatal-vs-log-panic/

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has the configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

File: client/pd_service_discovery_test.go [hunk 1]
```diff
@@ -107,10 +107,10 @@ func newTestServer(isLeader bool) *testServer {
 func (s *testServer) run() {
 	lis, err := net.Listen("tcp", s.addr)
 	if err != nil {
-		log.Fatalf("failed to serve: %v", err)
+		log.Panicf("failed to serve: %v", err)
```
Comment (HuSharp) [inline client/pd_service_discovery_test.go:110]: maybe can `log.Panic(zap.error())` ?

**Review by HuSharp — APPROVED** [review]

File: cmd/pd-server/main.go [hunk 2]
```diff
@@ -254,7 +254,7 @@ func start(cmd *cobra.Command, args []string, services ...string) {
 
 	err = join.PrepareJoinCluster(cfg)
 	if err != nil {
-		log.Fatal("join meet error", errs.ZapError(err))
+		log.Panic("join meet error", errs.ZapError(err))
```
Comment (rleungx) [inline cmd/pd-server/main.go:257]: I have some questions:
1. The behavior is different from the previous 
2. Will it be printed out if we set the log level to fatal?
3. What if the panic has a recover hook?

Comment (okJiang) [reply cmd/pd-server/main.go:257]: 1. Yes
2. Yes, the log will be printed when log level is fatal
3. If exist a recover hook, the panic will be caught.

> from: https://www.quasilyte.dev/blog/post/log-fatal-vs-log-panic/
> The log.Panic vs log.Fatal is essentially panic vs os.Exit(1). The latter will not let deferred calls to run.

**Comment by okJiang** [comment]
Since some users might rely on the "fatal" keyword to define alarm rules, we won't make any changes to this.

**Pull request closed without merging**